### ISSUE TITLE
nested @import fails when importer is async

### ISSUE THREAD

**Issue by nelisbijl** [medialize/sass.js#59]
When the importer does not call done synchronously, a nested @import fails.
See sample code below
If you comment out the async invocation and uncomment the direct invocation, things work as expected

```javascript
import Sass from 'sass.js';
var scssSuccess = '@import "child"; .parent { color:green; }',
    scssFail = '.parent { @import "child"; }';

Sass.importer((request, done) => {
    var childCss = '.child { color: red; }';

    var processContent = (content) => {
        var result = {
            content: content,
            path: 'file:///C:/TMP/child.scss'
        };
        console.log('importer\n  result', result);
        done(result);
    };

    // delayed (async) invocation of done
    new Promise(resolve => { resolve(childCss); }).then(processContent);

    // direct invocation of done
    //processContent(childCss);
});

async function compile(scss) {
    return await new Promise(resolve => {
        console.log('sass.compile\n  content', '>>>' + scss + '<<<');
        Sass.compile(scss, resolve);
    }).then(function (result) {
        console.log('result', result);
        return this;
    });
}

compile(scssSuccess).
then(() => {
    console.log('\n\nFAIL SITUATION');
    compile(scssFail);
});
```

**Comment by rodneyrehm** [comment]
Can you provide your code in a way that would run in a browser or Node (whatever version)? Or specify how to make your code run in either of those environments? Right now I'm getting syntax errors for `import` and `async`…

---

anyway, rewriting your example:

```js
var Sass = require('./dist/sass.sync.js');
var nestedImport = '.parent { @import "child"; }';

Sass.importer(function(request, done) {
  var importedContent = '.child { color: red; }';
  setTimeout(function() {
    var result = {
      content: importedContent,
      path: 'file:///C:/TMP/child.scss'
    };

    done(result);
  }, 100);
});

Sass.compile(nestedImport, function(result) {
  console.log(result);
});
```

yields the following error:

```
Error: Invalid CSS after '...@import "child"': expected "{", was "; }"
        on line 1 of stdin
>> .parent { @import "child"; }
   -------------------------^
```

But when the Importer callback returns immediately, things works out fine:

```js
var Sass = require('./dist/sass.sync.js');
var nestedImport = '.parent { @import "child"; }';

Sass.importer(function(request, done) {
  var importedContent = '.child { color: red; }';
  var result = {
    content: importedContent,
    path: 'file:///C:/TMP/child.scss'
  };

  done(result);
});

Sass.compile(nestedImport, function(result) {
  console.log(result);
});
```

produces the following `result.text`:

```css
.parent .child {
  color: red; }
```

Off the top of my head I've got no idea why this works synchronously, but fails asynchronously. The way [Importer callbacks are handled](https://github.com/medialize/sass.js/blob/master/src/sass.importer.js#L19-L31) in JS doesn't look like it could cause this. But [sass_importer_emscripten()](https://github.com/medialize/sass.js/blob/master/libsass/emscripten_wrapper.cpp#L122-L194) doesn't explain the behavior, either.

**Comment by nelisbijl** [comment]
I am using jspm, system.js, sass-plugin, es6 and babel. Hence the code sample. Your code produces the same result.
It seems like the finished method is called only once even if the importer hasn't finished.
The call stack however doesn't help much due to the emscripten code.
Any pointers, ideas how to fix this?

**Comment by rodneyrehm** [comment]
> It seems like the finished method is called only once even if the importer hasn't finished.

That's indeed what is happening. But since the returned value is "not done", [emscripten should sleep for 20ms and then try again](https://github.com/medialize/sass.js/blob/master/libsass/emscripten_wrapper.cpp#L151-L161). But that's not happening. It falls through on the first attempt. I suspect this is because `emscripten_sleep()` somehow breaks the loop.

> Any pointers, ideas how to fix this?

If I were to guess, I'd say the nested import is has a function in its call stack, that is not included in the [whitelist](https://github.com/medialize/sass.js/blob/master/libsass/empterpreter_whitelist.json). You'd have to run a debug build to get to this info… Not much fun if you haven't got things configured already…

I'll try to have a look at this on Sunday (as I'm currently on vacation and the lady is pestering me about going to Dinner…).

**Comment by rodneyrehm** [comment]
fixed the problem and released [0.9.11](https://github.com/medialize/sass.js/releases/tag/0.9.11) - thanks for filing the issue :)

**Comment by nelisbijl** [comment]
Thanks Rodney.

Works like a charm!
